### PR TITLE
`rename_all` container attribute to have all children fit a naming convention

### PR DESCRIPTION
**Pull request by 46bit**
Hi! This implements the `rename_all` crystallised at the bottom of #140.

* On an enum it renames the variants.
* On a struct variant it renames the variant's fields.
* On a struct it renames the fields.

There are 5 rename rules defined, all of them to change the naming convention:

* `#[serde(rename_all = "PascalCase")]`
* `#[serde(rename_all = "camelCase")]`
* `#[serde(rename_all = "snake_case")]`
* `#[serde(rename_all = "SCREAMING_SNAKE_CASE")]`
* `#[serde(rename_all = "kebab-case")]`

The names very visibly convey the relevant convention. I used the [Inflector](https://github.com/whatisinternet/inflector) crate to apply the naming convention.

### CODE REVIEW

File: serde_codegen_internals/src/attr.rs [hunk 1]
```diff
@@ -85,12 +87,57 @@ impl Name {
     }
 }
 
+#[derive(Debug, PartialEq)]
+pub enum RenameRule {
+    /// Don't apply a default rename rule.
+    None,
+    /// Rename direct children to "PascalCase" style, as typically used for enum variants.
+    PascalCase,
+    /// Rename direct children to "kebabCase" style.
```
Comment (dtolnay) [inline serde_codegen_internals/src/attr.rs]: this should say "camelCase"

File: serde_codegen_internals/src/attr.rs [hunk 2]
```diff
@@ -297,21 +360,25 @@ impl Item {
...
+                    content: content,
+                }
             }
             (true, Some(_), Some(_)) => {
                 cx.error("untagged enum cannot have #[serde(tag = \"...\", content = \"...\")]");
                 EnumTag::External
             }
         };
-
+        // @DEBUG
```
Comment (dtolnay) [inline serde_codegen_internals/src/attr.rs]: Has this been debugged? It looks all right to me.

Comment (46bit) [inline serde_codegen_internals/src/attr.rs]: Derp. Sorry. This is old.

**Comment by dtolnay** [comment]
This is very nicely done. The only major thing is as we discussed in IRC - let's figure out a way to avoid the heavyweight inflector dependency, whether that means grabbing the tiny amount of their code that we actually need, re-implementing some of it (each of these conversions is probably something like 10 lines), or making a PR to inflector to put some of the heavier functionality like pluralize and singularize behind a cfg.

**Comment by 46bit** [comment]
(Leaving here for a reminder.)

It might be best to implement very simple conversions, since we know from Rust conventions what the input formatting should be. `RenameRule::apply_to_variant` and `RenameRule::apply_to_field`.

**Comment by whatisinternet** [comment]
@46bit I've added a little spike and PR for feature flagging Inflector https://github.com/whatisinternet/inflector/pull/44. It's not 100% final, but I think it's something I can get on board with. Let me know if there is anything more I can do to help out. I'd appreciate a code review if/when you have time.

**Comment by 46bit** [comment]
@whatisinternet Sounds great! I'd definitely prefer using your library to my own conversions and this will enable that :-)

I just wonder whether `without_full` is the best name long-term. Breaking each module into a separate feature is a bit excessive but maybe something like `lightweight` better conveys the spirit of this?

**Comment by whatisinternet** [comment]
Agreed (names are hard). I've changed it to be `lightweight` which does feel better. WRT breaking each module out into a feature: It does feel a bit excessive at the moment. Let me know how the testing goes with the current feature gate, and we can reevaluate if needed.

**Comment by alexbool** [comment]
Perhaps @dtolnay should make a release with this magnificent feature!

**Comment by dtolnay** [comment]
I published this magnificent feature in Serde 0.9.10. Thanks for the reminder.

**Pull request merged**